### PR TITLE
Feature/hp mab floating devices

### PR DESCRIPTION
**Pull request by julsemaan**
# Description

Adds MAB floating devices support to HP Procurve switches
# Impacts

HP Procurve switches
# Delete branch after merge

YES
# NEWS file entries
## New Features
- Added MAB floating devices support to HP Procurve switches

### CODE REVIEW

File: lib/pf/Switch/HP/Procurve_2500.pm [hunk 1]
```diff
@@ -37,6 +39,120 @@ sub supportsWiredMacAuth { return $TRUE; }
...
+    return 1;
+}
+
+=head2 setTaggedVlans
+
+Tag a list of VLANs on a port
+
+=cut
+
+sub setTaggedVlans {
```
Comment (fdurand) [inline lib/pf/Switch/HP/Procurve_2500.pm:137]: Don't you have to create a removeAllTaggedVlans funtion ?

Comment (julsemaan) [inline lib/pf/Switch/HP/Procurve_2500.pm:137]: Nothing like that is called in floatingdevice.pm

And removing all VLANs that are tagged would mean we would be removing the voice VLAN.

I tested and leaving the tagged VLANs on the interface doesn't interfere with MAB once it is reactivated (even if the device ends up being in a VLAN defined as tagged)

**Comment by erSitzt** [comment]
On ProCurve 2610,2620 and 2920 i use this OID to turn port-access mac-based on and off

1.3.6.1.4.1.11.2.14.11.5.1.19.2.1.1.3
A value of 1 is enabled , a value of 2 is disabled

Maybe this works for all HP ProCurve ?

**Comment by erSitzt** [comment]
This is what i was testing with...

```
sub disableMABByIfIndex {
    my ( $this, $ifIndex ) = @_;
    my $logger = Log::Log4perl::get_logger(__PACKAGE__);

    if ( !$this->isProductionMode() ) {
        $logger->warn("Should set cafPortAuthorizeControl on $ifIndex to 3:forceAuthorized but the s");
        return 1;
    }

    if ( !$this->connectWrite() ) {
        return 0;
    }
# ProCurve port-access mac-based 1.3.6.1.4.1.11.2.14.11.5.1.19.2.1.1.3
    my $OID_hpicfUsrAuthMacAuthAdminStatus = '1.3.6.1.4.1.11.2.14.11.5.1.19.2.1.1.3';

    $logger->trace("SNMP set_request for hpicfUsrAuthMacAuthAdminStatus: $OID_hpicfUsrAuthMacAuthAdminStatus");
    my $result = $this->{_sessionWrite}->set_request(
        -varbindlist => [ "$OID_hpicfUsrAuthMacAuthAdminStatus.$ifIndex", Net::SNMP::INTEGER, 2 ] );
    return ( defined($result) );
}

sub enableMABByIfIndex {
    my ( $this, $ifIndex ) = @_;
    my $logger = Log::Log4perl::get_logger(__PACKAGE__);

    if ( !$this->isProductionMode() ) {
        $logger->warn("Should set hpicfUsrAuthMacAuthAdminStatus on $ifIndex to 2:auto but the switch is no");
        return 1;
    }

    if ( !$this->connectWrite() ) {
        return 0;
    }

    my $OID_hpicfUsrAuthMacAuthAdminStatus = '1.3.6.1.4.1.11.2.14.11.5.1.19.2.1.1.3';

    $logger->trace("SNMP set_request for hpicfUsrAuthMacAuthAdminStatus: $OID_hpicfUsrAuthMacAuthAdminStatus");
    my $result = $this->{_sessionWrite}->set_request(
        -varbindlist => [ "$OID_hpicfUsrAuthMacAuthAdminStatus.$ifIndex", Net::SNMP::INTEGER, 1 ] );
    return ( defined($result) );
}
```

The logger messages and some other stuff are not correct, because i reused stuff from other switchmodules... as i said this was just a test ;)

**Comment by julsemaan** [comment]
Looks promising.

What firmware version are you running ?

We can try SNMP instead of SSH since its cleaner

**Comment by julsemaan** [comment]
@erSitzt 

Might as well test if the VLAN tagging OIDs are supported

Can you try to walk these tables : 

```
    my $OID_vlanTrunkPortVlansEnabled   = '1.3.6.1.4.1.9.9.46.1.6.1.1.4';
    my $OID_vlanTrunkPortVlansEnabled2k = '1.3.6.1.4.1.9.9.46.1.6.1.1.17';
    my $OID_vlanTrunkPortVlansEnabled3k = '1.3.6.1.4.1.9.9.46.1.6.1.1.18';
    my $OID_vlanTrunkPortVlansEnabled4k = '1.3.6.1.4.1.9.9.46.1.6.1.1.19';
```

**Comment by erSitzt** [comment]
Those do not work for 2610,2620 and 2920

But i was using this (again very experimental ;) )

```
sub setTagVlansByIfIndex {
    my ( $this, $ifIndex, $setTo, $switch_locker_ref, @vlans ) = @_;
    my $logger = Log::Log4perl::get_logger( ref($this) );
    my $OID_rcVlanPortMembers = '1.3.6.1.2.1.17.7.1.4.3.1.2'; #RC-VLAN-MIB
$logger->info("Interface : $ifIndex - SetTo : $setTo");
    if ( !$this->connectRead() ) {
        return 0;
    }
    if ( !$this->connectWrite() ) {
        return 0;
    }

    my $result;
    {
        #lock %{ $switch_locker_ref->{ $this->{_ip} } };

        # since all VLANs are located in separated OIDs we need to fetch the portList for each VLAN
        my $fetch_vlan_port_list_ref = [];
        foreach my $vlan (@vlans) {
            push @$fetch_vlan_port_list_ref, "$OID_rcVlanPortMembers.$vlan";
        }

        $logger->info("SNMP get_request for rcVlanPortMembers");
        $this->{_sessionRead}->translate(0);
        $result = $this->{_sessionRead}->get_request( -varbindlist => $fetch_vlan_port_list_ref );
        $this->{_sessionRead}->translate(1);
#$logger->info(Dumper($result));
        # now we traverse every portList and set the proper port bit to 1 for every VLAN
        my $set_vlan_port_list_ref = [];
        foreach my $vlan (@vlans) {
            my $updated_port_member_list = $this->modifyBitmask(
                $result->{"$OID_rcVlanPortMembers.$vlan"}, $ifIndex - 1, $setTo
            );

            push @$set_vlan_port_list_ref,
                "$OID_rcVlanPortMembers.$vlan", Net::SNMP::OCTET_STRING, $updated_port_member_list;
        }
        $logger->info(Dumper($set_vlan_port_list_ref));
        $logger->trace( "SNMP set_request for OID_rcVlanPortMembers: $OID_rcVlanPortMembers");
        $result = $this->{_sessionWrite}->set_request(-varbindlist => $set_vlan_port_list_ref);
    }
    $logger->trace( "locking - \$switch_locker{" . $this->{_ip} . "} unlocked in setTaggedVlan" );

    # if $result is defined, it works we can return $TRUE
    return $TRUE if (defined($result));

    # otherwise report failure
    $logger->warn("Tagging VLANs failed with " . $this->{_sessionWrite}->error());
    return;
}
```

**Comment by erSitzt** [comment]
Tagging works with this, my only problem when i had to stop testing for a while was that disconnects of the floating device were not handled... but at that point i did not understand the whole process :) i still dont...

Here is the counterpart

```
sub removeAllTaggedVlans {
    my ( $this, $ifIndex, $switch_locker_ref ) = @_;

    my @all_vlans = keys %{$this->getVlans()};
    return $this->setTagVlansByIfIndex($ifIndex, $FALSE, $switch_locker_ref, @all_vlans);
}
```

**Comment by julsemaan** [comment]
Disconnects are handled by an SNMP up trap and should be reactivating the MAB and removing the trunk (in the case of HP there is no trunk).

By default it doesn't untag VLANs in our current flow as we can't be sure which VLANs to untag since the voice VLAN will be one of them on VoIP switches.

I will give the SNMP a try and update this PR if it is successful.

**Comment by erSitzt** [comment]
Nice !

This is the OID for port-security btw. 
1.3.6.1.4.1.11.2.14.2.10.3.1.6.1

I wrote this before learning that port-security should not be used anymore ( am i right here? ), but maybe it is still usefull.

```
sub isPortSecurityEnabled {
    my ( $this, $ifIndex ) = @_;
    my $logger = Log::Log4perl::get_logger( ref($this) );

    #CISCO-PORT-SECURITY-MIB
    my $OID_cpsIfPortSecurityEnable = '1.3.6.1.4.1.11.2.14.2.10.3.1.6.1';

    if ( !$this->connectRead() ) {
        return 0;
    }

    #determine if port security is enabled
    $logger->info(
        "SNMP get_request for cpsIfPortSecurityEnable: $OID_cpsIfPortSecurityEnable.$ifIndex"
    );
    my $result = $this->{_sessionRead}->get_request(
        -varbindlist => [ "$OID_cpsIfPortSecurityEnable.$ifIndex" ] );
    return (
        exists( $result->{"$OID_cpsIfPortSecurityEnable.$ifIndex"} )
            && ( $result->{"$OID_cpsIfPortSecurityEnable.$ifIndex"} ne
            'noSuchInstance' )
            && ( $result->{"$OID_cpsIfPortSecurityEnable.$ifIndex"} ne
            'noSuchObject' )
            && ( $result->{"$OID_cpsIfPortSecurityEnable.$ifIndex"} == 2 )
    );
}

=item setPortSecurityEnableByIfIndex - enable/disable port-security on a port

=cut

sub setPortSecurityEnableByIfIndex {
    my ( $this, $ifIndex, $enable ) = @_;
    my $logger = Log::Log4perl::get_logger( ref($this) );
$logger->info("Disable Port-Security");
    if ( !$this->isProductionMode() ) {
        $logger->warn("Should set IfPortSecurityEnable on $ifIndex to $enable but the switch is not in production -> Do nothing");
        return 1;
    }

    if ( !$this->connectWrite() ) {
        return 0;
    }
    my $OID_cpsIfPortSecurityEnable = '1.3.6.1.4.1.11.2.14.2.10.3.1.6.1';
$logger->info("DENNIS :: True : $SNMP::TRUE / False : $SNMP::FALSE");
    my $truthValue = $enable ? $SNMP::TRUE : $SNMP::FALSE;

    $logger->info("SNMP set_request for cpsIfPortSecurityEnable: $OID_cpsIfPortSecurityEnable");
    my $result = $this->{_sessionWrite}->set_request(
        -varbindlist => [ "$OID_cpsIfPortSecurityEnable.$ifIndex", Net::SNMP::INTEGER, $truthValue ] );
    return ( defined($result) );
}
```

**Comment by julsemaan** [comment]
You are right, port-security is in the deprecation phase (both in PacketFence and the network vendors) and its support is extremely likely to be dropped in PF6

**Comment by julsemaan** [comment]
@erSitzt 

Can you try the proposed switch module on your end and see if it works.

It should have all it needs to support floating devices.

I will also test it in lab

**Comment by erSitzt** [comment]
I'm currently on my way home, i'll test it tomorrow.

**Comment by erSitzt** [comment]
I'm not sure if this is needed to avoid errors, but i have this in my test switch module

```
sub isTrunkPort {
    my ( $this, $ifIndex ) = @_;
    my $logger = Log::Log4perl::get_logger( ref($this) );

    $logger->info("Trunk Port not necessary on ProCurve. Switch type: " . ref($this));
    return ( 1 );
}

sub setModeTrunk {
    my ( $this, $ifIndex ) = @_;
    my $logger = Log::Log4perl::get_logger( ref($this) );

    $logger->info("Function not needed for switch type " . ref($this));
    return ( 1 );
}
```

**Comment by erSitzt** [comment]
There seems to be a problem with tagging the vlans (8,15,16 are configured for the floating device), but the first entry in @vlans is undef

```
Dec 09 13:33:05 httpd.aaa(26160) INFO: Disabling port access control on port 1 (pf::floatingdevice::enablePortConfig)
Dec 09 13:33:05 httpd.aaa(26160) INFO: Disabling MAB on 1 (pf::Switch::HP::Procurve_2500::disableMABByIfIndex)
Dec 09 13:33:06 httpd.aaa(26160) INFO: Setting port 1 as trunk. (pf::Switch::enablePortConfigAsTrunk)
Dec 09 13:33:06 httpd.aaa(26160) INFO: Function not needed for switch type pf::Switch::HP::Procurve_2500 (pf::Switch::HP::Procurve_2500::setModeTrunk)
Dec 09 13:33:06 httpd.aaa(26160) INFO: Allowing tagged Vlans on port 1 (pf::Switch::enablePortConfigAsTrunk)
Dec 09 13:33:06 httpd.aaa(26160) INFO: $VAR1 = undef;
 (pf::Switch::HP::Procurve_2500::setTaggedVlans)
Dec 09 13:33:06 httpd.aaa(26160) INFO: $VAR1 = '8';
 (pf::Switch::HP::Procurve_2500::setTaggedVlans)
Dec 09 13:33:06 httpd.aaa(26160) INFO: $VAR1 = '15';
 (pf::Switch::HP::Procurve_2500::setTaggedVlans)
Dec 09 13:33:06 httpd.aaa(26160) INFO: $VAR1 = '16';
 (pf::Switch::HP::Procurve_2500::setTaggedVlans)
Dec 09 13:33:06 httpd.aaa(26160) INFO: $VAR1 = [
          '1.3.6.1.2.1.17.7.1.4.3.1.2.',
          4,
          '�',
          '1.3.6.1.2.1.17.7.1.4.3.1.2.8',
          4,
          '�',
          '1.3.6.1.2.1.17.7.1.4.3.1.2.15',
          4,
          '�',
          '1.3.6.1.2.1.17.7.1.4.3.1.2.16',
          4,
          '�'
        ];
 (pf::Switch::HP::Procurve_2500::setTaggedVlans)
Dec 09 13:33:06 httpd.aaa(26160) WARN: Tagging VLANs failed with The OBJECT IDENTIFIER value "1.3.6.1.2.1.17.7.1.4.3.1.2." is expected in dotted decimal notation (pf::Switch::HP::Procurve_2500::setTaggedVlans)
Dec 09 13:33:06 httpd.aaa(26160) ERROR: An error occured while allowing tagged Vlans on trunk port 1 (pf::Switch::enablePortConfigAsTrunk)
Dec 09 13:33:06 httpd.aaa(26160) INFO: Setting PVID as 1 on port 1. (pf::floatingdevice::enablePortConfig)
Dec 09 13:33:06 httpd.aaa(26160) INFO: There is a floating device on 172.16.4.17 port 1 (pf::floatingdevice::portHasFloatingDevice)
Dec 09 13:33:06 httpd.aaa(26160) INFO: Not adding locationlog entry for mac f0:9c:e9:34:c3:00 because it's plugged in a floating device enabled port (pf::locationlog::locationlog_synchronize)
Dec 09 13:33:06 httpd.aaa(26160) INFO: setting VLAN at 172.16.4.17 ifIndex 1 from 127 to 1 (pf::Switch::setVlan)
```

But i'm still running an older version of PacketFence, so this might be related to my setup.

**Comment by julsemaan** [comment]
On my end, the same code produced an error when trying to modify the bit mask since I believe the VLAN tagging OID didn't exist on the switch I was testing with.

I will see how we can address that. Otherwise I guess it will have to be SSH

**Comment by erSitzt** [comment]
Ok, i will try to finish my stuff for the 2610, 2620 and 2920 and create a pull request for the changes.
...after i updated my installation...

**Comment by julsemaan** [comment]
Reverted to using SSH for now since it should work on a Procurve since the commands are the same on all of them

**Comment by louismunro** [comment]
Good enough.
Merging.

**Pull request merged**